File: leetcode/medium/0003-longest-substring-without-repeating-characters/solution.go
func lengthOfLongestSubstring(s string) int {
    charIndex := make(map[byte]int)
    left := 0
    maxLength := 0

    for right := 0; right < len(s); right++ {
        if index, found := charIndex[s[right]]; found && index >= left {
            left = index + 1
        }

        charIndex[s[right]] = right

        if right-left+1 > maxLength {
            maxLength = right - left + 1
        }
    }

    return maxLength
}

File: leetcode/medium/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_index = {}
        left = 0
        max_length = 0

        for right, ch in enumerate(s):
            if ch in char_index and char_index[ch] >= left:
                left = char_index[ch] + 1

            char_index[ch] = right
            max_length = max(max_length, right - left + 1)

        return max_length